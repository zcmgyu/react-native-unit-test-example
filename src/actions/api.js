import { asyncActionNames, buildAsyncActions } from './utils/asyncUtils.js';

// Constants
export const LOAD_REPOS = asyncActionNames('LOAD_REPOS');

// Actions
export const receiveRepos = buildAsyncActions(LOAD_REPOS);

// Async actions
export const fetchData = (endpoint, fetchAction) => async dispatch => {
  try {
    dispatch(fetchAction.request());
    const response = await fetch(endpoint);
    const responseJson = await response.json();
    dispatch(fetchAction.success(responseJson.items));
  } catch (error) {
    dispatch(fetchAction.failure(error));
  }
};
