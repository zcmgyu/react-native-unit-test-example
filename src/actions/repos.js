// Constants
export const SELECT_REPO = 'SELECT_REPO';

// Actions
export const selectRepo = id => {
  return {
    type: SELECT_REPO,
    id
  };
};
