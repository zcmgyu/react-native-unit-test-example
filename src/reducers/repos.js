import { LOAD_REPOS } from '../actions/api';
import { SELECT_REPO } from '../actions/repos';

export const initialState = {
  // Exporting it for test purposes
  error: null,
  isLoading: false,
  repos: null,
  selected: null
};

export default function repos(state = initialState, action) {
  switch (action.type) {
    case LOAD_REPOS.failure:
      return { ...state, error: action.error, isLoading: false };
    case LOAD_REPOS.request:
      return { ...state, isLoading: true };
    case LOAD_REPOS.success:
      return { ...state, isLoading: false, repos: action.data };
    case SELECT_REPO:
      const selected = state.selected !== action.id ? action.id : null;
      return { ...state, selected };
    default:
      return state;
  }
}
