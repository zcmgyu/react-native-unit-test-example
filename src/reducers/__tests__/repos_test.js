import repos from '../repos';
import { initialState } from '../repos';
import { receiveRepos } from '../../actions/api';
import { selectRepo } from '../../actions/repos';
import data from '../../../config/jest/dataMock';


it('returns the same state on an unhandled action', () => {
  expect(repos(initialState, { type: 'HELLO_WORLD' })).toMatchSnapshot();
});

it('handles LOAD_REPOS_REQUEST action', () => {
  expect(repos(initialState, receiveRepos.request())).toMatchSnapshot();
});

it('handles LOAD_REPOS_FAILURE action', () => {
  const error = new Error('Look ma! I am an error');
  expect(repos(initialState, receiveRepos.failure(error))).toMatchSnapshot();
});

it('handles LOAD_REPOS_SUCCESS action', () => {
  // BOOM. If we change that specific reducer case, we'll compare the Snapshot results
  // and update them if the change is expected.
  expect(repos(initialState, receiveRepos.success(data))).toMatchSnapshot();
});

it('handles SELECT_REPO action', () => {
  const newState = repos(initialState, selectRepo(1));
  expect(newState).toMatchSnapshot();
  // Sending the same id should clean it
  expect(repos(newState, selectRepo(1))).toMatchSnapshot();
});