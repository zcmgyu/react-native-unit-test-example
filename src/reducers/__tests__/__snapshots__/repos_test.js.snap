// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles LOAD_REPOS_FAILURE action 1`] = `
Object {
  "error": [Error: Look ma! I am an error],
  "isLoading": false,
  "repos": null,
  "selected": null,
}
`;

exports[`handles LOAD_REPOS_REQUEST action 1`] = `
Object {
  "error": null,
  "isLoading": true,
  "repos": null,
  "selected": null,
}
`;

exports[`handles LOAD_REPOS_SUCCESS action 1`] = `
Object {
  "error": null,
  "isLoading": false,
  "repos": Array [
    Object {
      "description": "Useful description goes here.",
      "id": 1,
      "name": "Test 1",
      "stargazers_count": 50000,
    },
    Object {
      "description": "Useful description goes here.",
      "id": 2,
      "name": "Test 2",
      "stargazers_count": 10000,
    },
    Object {
      "description": "Useful description goes here.",
      "id": 3,
      "name": "Test 3",
      "stargazers_count": 499,
    },
  ],
  "selected": null,
}
`;

exports[`handles SELECT_REPO action 1`] = `
Object {
  "error": null,
  "isLoading": false,
  "repos": null,
  "selected": 1,
}
`;

exports[`handles SELECT_REPO action 2`] = `
Object {
  "error": null,
  "isLoading": false,
  "repos": null,
  "selected": null,
}
`;

exports[`returns the same state on an unhandled action 1`] = `
Object {
  "error": null,
  "isLoading": false,
  "repos": null,
  "selected": null,
}
`;
